format/ComputeIR.fbs

```diff
@@ -0,0 +1,348 @@
+  /// up child fields, dictionary encoding, etc.
+  field: Field (required);
+
+  /// Buffers containing N elements of arrow-formatted data, where N
+  /// is Array.length if shape is Array or 1 if shape is Scalar.
+  /// XXX this can be optimized for trivial scalars later
+  buffers: [InlineBuffer];
+
+  /// If (and only if) this Literal has dictionary type, this field dictionary
+  /// into which the literal's indices refer.
+  dictionary: Literal;
+}
+
+table FieldRef {
+  /// A sequence of field names to allow referencing potentially nested fields
+  path: [string];
+
+  /// For Expressions which might reference fields in multiple Relations,
+  /// this index may be provided to indicate which Relation's fields
+  /// `path` points into. For example in the case of a join,
+  /// 0 refers to the left relation and 1 to the right relation.
+  relation_index: int;
+
+  /// The type of the referenced Field. Field is used instead of Type to pick
+  /// up child fields, dictionary encoding, etc.
+  field: Field;
+}
+
+/// A canonical (probably SQL equivalent) function
+enum CanonicalFunctionId : uint32 {
+  // logical
+  And,
+  Not,
+  Or,
+
+  // arithmetic
+  Add,
+  Subtract,
+  Multiply,
+  Divide,
+  Power,
+  AbsoluteValue,
+  Negate,
+  Sign,
+
+  // comparison
+  Equal,
+  NotEqual,
+  Greater,
+  GreaterOrEqual,
+  Less,
+  LessOrEqual,
+
+  // aggregations
+  All,
+  Any,
+  Count,
+  Mean,
+  Min,
+  Max,
+  Mode,
+  Product,
+  Sum,
+  Tdigest,
+  Quantile,
+  Variance,
+  StandardDeviation,
+}
+
+table CanonicalFunction {
+  id: CanonicalFunctionId;
+}
+
+table NonCanonicalFunction {
+  name_space: string (required);
+  name: string (required);
+}
+
+union Function {
+  CanonicalFunction, NonCanonicalFunction
+}
+
+table Call {
+  /// The function whose invocation this Call represents.
+  function: Function (required);
+
+  /// Parameters for `function_name`; content/format may be unique to each
+  /// value of `function_name`.
+  options: InlineBuffer;
+
+  /// The arguments passed to `function_name`.
+  arguments: [Expression] (required);
+
+  /// The type of data which invoking `function_name` will return.
+  /// Field is used instead of Type to pick up child fields,
+  /// dictionary encoding, etc.
+  field: Field;
+}
+
+enum CanonicalOperationId : uint32 {
+  Literal,
+  Filter,
+  Project,
+  Aggregate,
+  Join,
+  OrderBy,
+  Limit,
+  Common,
+  Union,
+  InteractiveOutput,
+}
+
+table CanonicalOperation {
+  id: CanonicalOperationId;
+}
+
+table NonCanonicalOperation {
+  name_space: string (required);
+  name: string (required);
+}
+
+union Operation {
+  CanonicalOperation, NonCanonicalOperation
+}
+
+/// A relation is a set of rows with consistent schema.
+table Relation {
+  /// The operation which this Relation wraps.
+  operation: Operation (required);
+
+  /// Parameters for `operation`; content/format may be unique to each
+  /// value of `operation`.
+  options: InlineBuffer;
```

Comment:
Having the options as a nested / independent flatbuffer for all canonical operations still makes me squirm. What do you think about putting a union of options for the canonical types in `CanonicalOperation` so that dealing with the InlineBuffer for "built-ins" is not necessary?

Reply:
-1. This adds yet more special casing for the canonical operations and makes it harder to write generic pattern matching utilities while also giving us another discriminant to query and validate.

Reply:
Another pattern might be to make CanonicalOperationId "CanonicalOperation" which is a union of the options.  It doesn't fully get out of slightly harder pattern matching code, but it would eliminate a separate discrimant.

Reply:
I think there's a tradeoff here. Having two code paths for handling canonical versus non canonical increases complexity, since a consumer now has to handle canonical things in additional to still handling InlineBuffer. If we're going to allow InlineBuffer, I don't think there's a good reason not to have just a single way to deal with options.

Reply:
Since we don't yet have a very specific idea of how people are going to use this, I think we should leave this as InlineBuffers.